### ISSUE TITLE
[J4.x] Media fields in non-repeatable custom fields subform toggle display on save

### ISSUE THREAD

**Issue by crystalenka** [joomla/joomla-cms#36734]
Buckle up, folks, this is a weird one.

### Steps to reproduce the issue
1. Create a media custom field. Configuration doesn't matter.
2. Create another custom field, any type but media. Configuration doesn't matter.
3. Create a subform with those two fields. **Set the "Repeatable" toggle to no. (This is important!)**
4. Open an article with the subform field. Fill out the fields, or don't. It doesn't matter.
5. Save the article.

### Expected result

The media field still displays in the subform, and if any value was added to the media field, it remains.

### Actual result

The media field _disappears_.

Before Save:
<img width="1414" alt="Screen Shot 2022-01-18 at 19 29 41" src="https://user-images.githubusercontent.com/9141288/149988210-b23832d3-6a81-402e-ad9d-3b9f3f610636.png">

After Save:
<img width="1414" alt="Screen Shot 2022-01-18 at 19 29 48" src="https://user-images.githubusercontent.com/9141288/149988226-d65bf75a-d946-40ff-b4cb-4f6f0b60e821.png">

Additionally, if you save the article again, the media field will magically reappear (but empty of any value). This happens regardless of what value (if any) you put in the media field. The only time the media field does NOT reappear is if it is the only field in the subform, which is why I included step 2 above.

### System information (as much as possible)

PHP Version: 8.0.14
Joomla! 4.0.6

### Additional comments
It's really fun if you save the article so the subform disappears, go back to the subform configuration, add another media field, and go back to edit the article. The media fields switch on every save and are never shown at the same time. 😶

**Comment by brianteeman** [comment]
confirmed

**Comment by crystalenka** [comment]
Thanks Brian. Felt like I was going crazy for a while!

**Comment by brianteeman** [comment]
that;s why I quickly posted a confirmation. No idea on the solutions (yet)

**Comment by richard67** [comment]
Maybe something for @Fedik ?

**Comment by richard67** [comment]
I wish we always had such precise instructions on how to reproduce an issue.

**Comment by brianteeman** [comment]
Must be related to the subform NOT being repeatable as everything works as expected when it is repeatable

**Comment by crystalenka** [comment]
Yup, that's why I included the important note on that step. :)

**Comment by brianteeman** [comment]
About to start filming but I found the problem Maybe someone else will write the fix while I am filming

The record saved in #__filed_values when it is **not** a repeatable subform 

{"field3":{"imagefile":"","alt_text":""},"field4":""}

The record saved in #__filed_values when it is a repeatable subform 

{"**row0**":{"field3":{"imagefile":"","alt_text":""},"field4":""}}


Adding the "row0" container to the field value on the non-repeatable subform and it works

**Comment by crystalenka** [comment]
I think the answer is in `plugins/fields/subform/subform.php`, possibly rows 174-177 because that's where the plugin tries to account for the non repeatable subform. I have no idea how to debug it but I suspect something is not working as expected there when the subform has a media field? 

I don't know, I'm wayyyy out of my depth here.

**Comment by nikosdion** [comment]
I have figured it out. It's a “simple” matter of data format discrepancy. I will submit a PR for 4.1 very soon.

**Comment by nikosdion** [comment]
There you go. It should fairly simple to test.

BTW, neither Brian nor Crystal got the right reason. Since Crystal asked me how did I debug it: The only way to figure out what is going on was to go through a debugger and walk backwards from the subform rendering. After several false starts, I hit `Joomla\CMS\Form\Field\SubformField::getInput` where I observed that the `$forms[0]` would include the media subfield in one page load and not the next after saving, then back again after saving again. I then noticed that when we call loadSubFormData in that class `$this->value` once contained 2 fields (and media field was not shown!) once it contained one field (and the media field WAS shown). How curious! This let me to put a breakpoint in the `Joomla\CMS\Form\Field\AccessiblemediaField` to see what the actual bleepit bleeping bleep is going on. I saw the array data coming in and falling through to the second branch of the if-block which provided the much necessary a-ha moment.
